LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ExclusiveProcessor.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+#pragma warning disable CA1003 // Use generic event handler instances (suppressed for performance)
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public sealed class ExclusiveProcessor<T>
+    {
+        private readonly Func<IReadOnlyList<T>, T> scheduler;
+        private readonly IEqualityComparer<T> comparer;
+        private readonly SemaphoreSlim processingLock = new(1);
+        private readonly List<T> queue = new();
+
+        public event EventHandler<T>? Submitted;
+        public event EventHandler<(T InterruptedProcessor, T InterruptingProcessor)>? Interrupted;
+        public event EventHandler<T>? Processing;
+        public event EventHandler<(T Processor, IProcessingOutcome Outcome)>? Processed;
+
+        public ExclusiveProcessor() : this(ps => ps[0]) { }
+
+        public ExclusiveProcessor(Func<IReadOnlyList<T>, T> scheduler) : this(scheduler, null) { }
+
+        public ExclusiveProcessor(Func<IReadOnlyList<T>, T> scheduler, IEqualityComparer<T>? comparer)
+        {
+            this.scheduler = scheduler;
```

Comment:
for me it is super hard to read the code easily with the function callback for the scheduler. If I want to understand what schedulers we have implemented, VS can't help me. I need to find all references to the ctor, look at all overloads and end up at the places, where we pass in a function callback for the scheduler. Vs. using a proper interface, or even a virtual method on the processor to allow implement a different scheduler (not my favorite), both would allow to find me all implementations with a single query.

Reply:
Refactored the scheduler to an interface with 8a5c52594de7590c4d76ce3c58af6a71999b0eef. Please resolve if the new version is what you had in mind.